include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,414 @@
+            mpi::reduceMethods::Reduce());
+
+        // Calculate intensity on master
+        if (isMaster)
+        {
+            for (unsigned int i = 0; i < n_q; i++)
+                intensity_master[i] =
+                    (sumfcoskr_master[i] * sumfcoskr_master[i] +
+                        sumfsinkr_master[i] * sumfsinkr_master[i]) /
+                    np_master.x();
+
+            std::stringstream o_step;
+            o_step << currentStep;
+            writeIntensity(intensity_master,
+                "saxsOutput/" + filename_prefix + "_" + o_step.str() + ".dat");
+            writeLog(np_master, nmp_master,
+                "saxsOutput/" + filename_prefix + "_" + o_step.str() + ".log");
+        }
+    }
+
+    // perform all operations to get data from GPU to master
+    void collectDataGPUToMaster()
+    {
+        // collect data GPU -> CPU -> Master
+        copyIntensityDeviceToHost();
+        collectIntensityOnMaster();
+    }
+
+    /**
+     * This functions calls the SAXS kernel. It specifies how the
+     * calculation is parallelized.
+     **/
+    void calculateSAXS(uint32_t currentStep)
+    {
+        this->currentStep = currentStep;
+
+        DataConnector &dc = Environment<>::get().DataConnector();
+        auto particles =
+            dc.get<ParticlesType>(ParticlesType::FrameType::getName(), true);
+
+        // calculate the absolute position of the particles
+        const SubGrid<simDim> &subGrid = Environment<simDim>::get().SubGrid();
+        DataSpace<simDim> globalOffset(subGrid.getLocalDomain().offset);
+
+        constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<
+            pmacc::math::CT::volume<SuperCellSize>::type::value>::value;
+
+        // initialize variables with zero
+        sumfcoskr->getDeviceBuffer().setValue(0.0);
+        sumfsinkr->getDeviceBuffer().setValue(0.0);
+        np->getDeviceBuffer().setValue(0.0);
+        nmp->getDeviceBuffer().setValue(0.0);
+
+        // calculate q_step
+        q_step[0] = (q_max[0] - q_min[0]) / n_qx;
+        q_step[1] = (q_max[1] - q_min[1]) / n_qy;
+        q_step[2] = (q_max[2] - q_min[2]) / n_qz;
+
+        // PIC-like kernel call of the SAXS kernel
+        PMACC_KERNEL(
+            KernelSaxs< numWorkers >{}
+        )(
+            1,
+            numWorkers
+        )(
+            // Pointer to particles memory on the device
+            particles->getDeviceParticlesBox(),
+            // Pointer to memory of sumfcoskr & sumfsinkr on the device
+            sumfcoskr->getDeviceBuffer().getDataBox(),
+            sumfsinkr->getDeviceBuffer().getDataBox(),
+            np->getDeviceBuffer().getDataBox(),
+            nmp->getDeviceBuffer().getDataBox(),
+            globalOffset,
+            currentStep,
+            *cellDescription,
+            subGrid.getGlobalDomain().size,
+            q_min,
+            q_max,
+            q_step,
+            n_qx,
+            n_qy,
+            n_qz,
```

Comment:
This is purely subjective, so not necessary to take into account at all. Why are `n_qx, n_qy, n_qz` treated as 3 separate numbers, while `q_min`, `q_max` and `q_step` as vectors? For me it seems logical to also put the `n_...` variables into a vector here and in the kernel. Also, in case `n_q` is always a product of components, it can be computed in the kernel and does not need to be passed.